src/process/path.h

```diff
@@ -154,6 +199,25 @@ static __always_inline int write_path(struct pt_regs *ctx, cached_path_t *cached
         if (bpf_probe_read(&offset, sizeof(offset), dentry + name) < 0)
             goto NameError;
 
+        // Path filtering, only if the filter is active
+        if (cached_path->filter_state >= 0) {
+            val = NULL;
+            // We need to reset at least the whole string storage array each time we use the key
+            bpf_probe_read(&key, sizeof(filter_key_t), &blank_key);
+            key.current_state = cached_path->filter_state;
+            if (bpf_probe_read_str(&key.path_segment, MAX_PATH_SEG, offset) > 0) {
+                val = (filter_value_t *)bpf_map_lookup_elem(&filter_transitions, &key);
+            }
+            // One retry with "*" path element if not found
+            if (val == NULL) {
+                // The retry_key always uses the same path segment string so we don't need to reset
+                retry_key.current_state = cached_path->filter_state;
+                val = (filter_value_t *)bpf_map_lookup_elem(&filter_transitions, &retry_key);
+            } 
+            // Terminate the filter lookup if there was no value at either key
+            cached_path->filter_state = val ? val->next_state : -1;
```

Comment:
for the cases where we *want* to do the filtering, after this returns `-1` any future iteration/tail call isn't needed (afaict) so should we be exiting early here? Maybe return something out of `write_path` that indicates "stopped because got filtered out"?

Reply:
I did some experiments with early exiting but I wanted to verify the logic before trying to take extra shortcuts. I think in practice you don't end up getting stuck in the state machine like that very often before the root, so it might not save much.